src/training/graph_group.h

```diff
@@ -58,19 +65,32 @@ class AsyncGraphGroup : public GraphGroup {
         return;
 
       // @TODO read guard on parameters
-      std::lock_guard<std::mutex> guard(sync_);
-      oldParams->copyFrom(params_);
+      int pos = 0;
+      for (int idx = 0; idx < devices_.size(); idx++) {
+        std::lock_guard<std::mutex> guard( shardSync_[idx] );
+        oldParams->subtensor(pos , params_[idx]->size())->copyFrom(params_[idx]);
+        pos += shardSize_;
```

Comment:
Weird, my latest push got good results on azure machine actually.

you might want try this: https://github.com/amunmt/marian/pull/36/commits/27bf4e376940e089f076112e62ec328b4fff0cf7

this one is slow compared to non-sharding, but back then the non-sharding was buggy.
Maybe the speed matched now?